src/test/rgw/test_redis_queue.cc

```diff
@@ -0,0 +1,354 @@
+        std::string invalid_lock_cookie = "invalid_cookie";
+        res = rgw::redisqueue::locked_read(
+            conn, "test_queue", invalid_lock_cookie, read_res, yield);
+        ASSERT_NE(res, 0);
+        ASSERT_EQ(res, -EBUSY);
+        ASSERT_EQ(read_res, "");
+
+        // Try to read an expired lock
+        boost::asio::steady_timer timer(io, std::chrono::milliseconds(500));
+        timer.async_wait(yield);
+
+        res = rgw::redisqueue::locked_read(conn, "test_queue", lock_cookie,
+                                           read_res, yield);
+        ASSERT_NE(res, 0);
+        ASSERT_EQ(res, -ENOENT);
+        ASSERT_EQ(read_res, "");
+      },
+      [this](std::exception_ptr eptr) {
+        conn->cancel();
+        if (eptr) std::rethrow_exception(eptr);
+      });
+  io.run();
+}
+
+// Use a valid lock and ack_read
+TEST_F(RGWRedisQueueTest, AckRead) {
+  io.restart();
+  boost::asio::spawn(
+      io,
+      [this](boost::asio::yield_context yield) {
+        int res;
+        std::string read_res;
+        std::tuple<int, int> status;
+
+        res = rgw::redisqueue::queueStatus(conn, "test_queue", status, yield);
+        ASSERT_EQ(res, 0);
+
+        int initial_reserve = std::get<0>(status);
+        int initial_queue = std::get<1>(status);
+
+        res = rgw::redisqueue::reserve(conn, "test_queue", yield);
+        ASSERT_EQ(res, 0);
+
+        std::string data = R"({
+          "Records": [
+            {
+              "version": "0",
+              "region": "test-region"
+            },
+            {
+              "version": "1",
+              "region": "test-region"
+            }
+          ]
+        })";
+
+        res = rgw::redisqueue::commit(conn, "test_queue", data, yield);
+        ASSERT_EQ(res, 0);
+
+        res = rgw::redisqueue::queueStatus(conn, "test_queue", status, yield);
+        ASSERT_EQ(res, 0);
+        ASSERT_EQ(std::get<0>(status), initial_reserve);
+        ASSERT_EQ(std::get<1>(status), initial_queue + 1);
+
+        std::string lock_name = "lock:test_queue";
+        std::string lock_cookie = "mycookie";
+        int duration = 500;
+        int return_code =
+            rgw::redislock::lock(conn, lock_name, lock_cookie, duration, yield);
+        ASSERT_EQ(return_code, 0);
+
+        res = rgw::redisqueue::locked_read(conn, "test_queue", lock_cookie,
+                                           read_res, yield);
+        ASSERT_EQ(res, 0);
+        ASSERT_EQ(data, read_res);
+
+        res = rgw::redisqueue::ack_read(conn, "test_queue", "mycookie", yield);
+        ASSERT_EQ(res, 0);
+
+        res = rgw::redisqueue::queueStatus(conn, "test_queue", status, yield);
+        ASSERT_EQ(res, 0);
+        ASSERT_EQ(std::get<0>(status), initial_reserve);
+        ASSERT_EQ(std::get<1>(status), initial_queue);
+      },
+      [this](std::exception_ptr eptr) {
+        conn->cancel();
+        if (eptr) std::rethrow_exception(eptr);
+      });
+  io.run();
+}
```

Comment:
please add test where multiple clients try to reserve, abort and commit to the queue

Reply:
they don't need to run in a separate coroutine. you can do that sequencially, just make sure that there are multiple reservations, and some of them are committed and some aborted.

running multiple clients in their own coroutines is also goo, but it checks for race conditions and other errors.